crates/sui/src/client_commands.rs

```diff
@@ -946,20 +965,39 @@ impl SuiClientCommands {
                 let sender = context.try_get_object_owner(&opts.gas).await?;
                 let sender = sender.unwrap_or(context.active_address()?);
                 let client = context.get_client().await?;
+                let chain_id = client.read_api().get_chain_identifier().await.ok();
+
+                let build_config = resolve_lock_file_path(build_config, Some(&package_path))?;
+                let previous_id = sui_package_management::set_package_id(
+                    &package_path,
+                    build_config.clone().install_dir,
+                    chain_id.clone(),
+                    AccountAddress::ZERO,
+                );
                 let package_path =
                     package_path
                         .canonicalize()
                         .map_err(|e| SuiError::ModulePublishFailure {
                             error: format!("Failed to canonicalize package path: {}", e),
                         })?;
-                let (dependencies, compiled_modules, _, _) = compile_package(
+                let compile_result = compile_package(
                     client.read_api(),
                     build_config.clone(),
                     &package_path,
                     with_unpublished_dependencies,
                     skip_dependency_verification,
                 )
-                .await?;
+                .await;
+                // Restore original ID, then check result.
+                if let Ok(Some(previous_id)) = previous_id {
+                    let _ = sui_package_management::set_package_id(
+                        &package_path,
+                        build_config.clone().install_dir,
+                        chain_id,
+                        previous_id,
+                    );
+                }
+                let (dependencies, compiled_modules, _, _) = compile_result?;
```

Comment:
I don't _love_ the how setting/resetting the original-id here bulks up this code block, but it is straightforward to reason about so I didn't wrestle further with it. 

I couldn't find a way to avoid `clone` in some places and tried some things 🤷

Reply:
Come to think of it, if you are doing a publish, do you need to set the previous ID? Presumably it will get overwritten based on transaction effects below?

Reply:
Yes, need to reset because publishing can fail (insufficient gas, etc.) and will leave the state as `0x0` in that case. It gets overwritten only if publishing succeeds.